subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependenciesAttributesIntegrationTest.groovy

```diff
@@ -146,14 +143,325 @@ class DependenciesAttributesIntegrationTest extends AbstractModuleDependencyReso
         resolve.expectGraph {
             root(":", ":test:") {
                 module('org:test:1.0') {
+                    configuration = expectedVariant
                     variant(expectedVariant, expectedAttributes)
                 }
             }
         }
 
         where:
         attributeValue | expectedVariant | expectedAttributes
-        'c1'           | 'api'           | ['org.gradle.status': 'release', 'org.gradle.usage': 'java-api', custom: 'c1']
-        'c2'           | 'runtime'       | ['org.gradle.status': 'release', 'org.gradle.usage': 'java-runtime', custom: 'c2']
+        'c1'           | 'api'           | ['org.gradle.status': defaultStatus(), 'org.gradle.usage': 'java-api', custom: 'c1']
+        'c2'           | 'runtime'       | ['org.gradle.status': defaultStatus(), 'org.gradle.usage': 'java-runtime', custom: 'c2']
+    }
+
+    @RequiredFeatures(
+        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
+    )
+    def "Merges consumer configuration attributes with dependency attributes (simulates Java Library)"() {
+        given:
+        repository {
+            'org:test:1.0' {
+                variant('api') {
+                    attribute('custom', 'c1')
+                }
+                variant('runtime') {
+                    attribute('custom', 'c2')
+                }
+            }
+        }
+
+        buildFile << """
+            configurations.conf.attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, "java-api"))
+
+            dependencies {
+                conf('org:test:1.0') {
+                    attributes {
+                        attribute(CUSTOM_ATTRIBUTE, 'c1')
+                    }
+                }
+            }
+        """
+
+        when:
+        repositoryInteractions {
+            'org:test:1.0' {
+                expectResolve()
+            }
+        }
+        succeeds 'checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(":", ":test:") {
+                module('org:test:1.0') {
+                    configuration = 'api'
+                    variant('api', ['org.gradle.status': DependenciesAttributesIntegrationTest.defaultStatus(), 'org.gradle.usage': 'java-api', custom: 'c1'])
+                }
+            }
+        }
+    }
+
+    @RequiredFeatures(
+        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
+    )
+    def "Fails resolution because consumer configuration attributes and dependency attributes conflict (simulates Java Library)"() {
```

Comment:
I find these "simulates Java library" comments more confusing than helpful, because it doesn't really simulate that behavior. E.g. for a Java library, the runtime is a valid substitute for the api, so this test would not fail.

Reply:
You're right, I'll remove the comment, it was mostly for me :)